runners/direct-java/src/main/java/org/apache/beam/runners/direct/UnboundedReadEvaluatorFactory.java

```diff
@@ -198,13 +201,8 @@ private boolean startReader(
     private UnboundedSourceShard<OutputT, CheckpointMarkT> finishRead(
         UnboundedReader<OutputT> reader, UnboundedSourceShard<OutputT, CheckpointMarkT> shard)
         throws IOException {
-      final CheckpointMark oldMark = shard.getCheckpoint();
       @SuppressWarnings("unchecked")
       final CheckpointMarkT mark = (CheckpointMarkT) reader.getCheckpointMark();
-      if (oldMark != null) {
```

Comment:
I believe this finalization should stay; otherwise we won't ever finalize any checkpoint except the last one.